snap/get_version

```diff
@@ -0,0 +1,114 @@
+#!/bin/dash
+set -eu
+
+# Script to build the version to set to the snap.
+
+# When considering a tag, if starting with "<current branch name>-" it will have its prefix removed.
+# For instance:
+# * 0.1 -> 0.1
+# * msentraid-0.1 -> 0.1 on msentraid branch.
+
+# 1. If current commit is tagged, the version is directly the tag name.
+# 2. If current commit is not tagged, the version is:
+#    a. <last tag name on current branch>+<commit_sha> for main branch.
+#    b. <last tag name on current branch (not in main)>+<commit_sha>.<last_commit_merged_from_main> for other branches.
+#
+# Any of those version will be annoted with +dirty if there are local changes.
+
+# set_version will markup the version in the snapcraft.yaml file after amending it with a dirty markup if necessary.
+# $1: version: the version to set.
+set_version() {
+    version="${1}"
+
+    version=$(annotate_with_dirty "${version}")
+    craftctl set version="${version}"
+}
+
+# annotate_with_dirty may amend the version with a dirty markup if there are local changes.
+# $1: version: the version to annotate.
+annotate_with_dirty() {
+    version="${1}"
+
+    # check if current tree content is dirty.
+    if [ -n "$(git -C ${SNAPCRAFT_PART_SRC} status --porcelain)" ]; then
+        version="${version}+dirty"
+    fi
+
+    echo "${version}"
+}
+
+# last_tag_on_branch returns the last tag on that branch that is not in another provided branch.
+# $1: origin_branch: branch we start listing from.
+# $2: other_branch: the branch where we don’t want to tags to be reachable from.
+# If we don’t have another branch to compare with, it will just return the most recent tag.
+last_tag_on_branch() {
+    origin_branch="${1}"
+    other_branch="${2}"
+
+    all_tags_on_origin_branch=$(git -C ${SNAPCRAFT_PART_SRC} tag --sort=-committerdate --merged "${origin_branch}")
+    for tag in ${all_tags_on_origin_branch}; do
+        # Just return first commit if there is no branch to compare against the main branch.
+        if [ -z "${other_branch}" ]; then
+            tag=$(strip_branch_tag_prefix "${tag}" "${origin_branch}")
+            echo ${tag}
+            return
+        fi
+
+        # Skip the tags that are reachable from the other branch.
+        if git -C ${SNAPCRAFT_PART_SRC} merge-base --is-ancestor $(git -C ${SNAPCRAFT_PART_SRC} rev-list -n 1 "${tag}") "${other_branch}"; then
```

Comment:
better to split for error handling.

Reply:
agreed, for debugging, especially if we add -x, it will be easier handled in 2 calls.